openedx/xblock-lti-consumer
# [BB-5559] Adds support for external LTI configurations using openedx-filters

### Background

Presently the LTI Component XBlock supports integration of LTI Tools by adding the necessary configuration values directly in the XBlock settings. This means things like the launch URL, tool id for Version 1.1 and Login URL, public key and other parameters for version 1.3 need to be entered at every place the XBlock is used.

This causes friction for the course creators while creating and updating the course, as each instance of the XBlock needs to be edited individually.

### Description

This PR is based on the [ADR](https://github.com/openedx/xblock-lti-consumer/blob/master/docs/decisions/0006-pluggable-lti-configuration.rst) aimed at reducing the friction by providing a way for the XBlock to fetch the necessary information for an LTI tool from an external location. 

The PR introduces the following changes:

1. This is achieved by using the events API provided by [openedx-filters](https://github.com/openedx/openedx-filters). We introduce a new filter in the XBlock: ```org.openedx.xblock.lti_consumer.configuration.listed.v1```
    -  As a proof of concept, a new plugin is also created called [openedx_ltistore](https://github.com/open-craft/openedx-ltistore/pull/1) which implements storing of the LTI configurations in the DB and the filter pipeline to return configurations when the filter is run has also been created.
 2. Adds an extra option `XBLOCK_EXTERNAL` to the `LtiConfiguration` model's `config_store` attribute.
 3. Adds an extra option `External Configuration` to the `LTI Version` select in the **Edit** form of the XBlock, which will allow the user to select an external configuration instead of specifying the version and entering related information.
 4. All of the changes are gated behind a Course level Waffle Flag - `lti_consumer.enable_external_config_filter`


### Testing Instructions

The testing instructions you are using the **openedx_devstack** to test.

1. Setup the [openedx_ltistore](https://github.com/open-craft/openedx-ltistore/pull/1) plugin as described in the README and add a couple of configurations with version 1.1. 
    - **Note:** [Saltire](https://saltire.lti.app/) can be used to get some configuration for testing
2. Create a sample configuration with LTI version 1.1 in the [LTI_Store admin|http://localhost:18010/admin/lti_store/] note down the **Filter Key** value of the newly created configuration.We will use this later. eg., `lti_store:1`
3. Setup `xblock-lti-consumer` in the development mode
4. Checkout to this PR branch
5. Restart studio using `make studio-restart`
6. Create a test course, if you don't already have one.
7. Enable the external config feature by creating a course waffle flag in the [admin page](http://localhost:18010/admin/waffle_utils/waffleflagcourseoverridemodel/)
    - FLAG:  `lti_consumer.enable_external_config_filter`
    - Course: `<your course id>`
    - Force: ON
    - Enabled
9. Add `lti_consumer` to the **Advanced Module List** in the **Advanced Settings** of the test course.
10. Add a new unit to the course and add the **LTI Consumer** from the **Advanced** blocks.
11. Click the **Edit** button of the LTI Consumer block and set **Configuration Type** to `Existing Configuration`
12. Set the **External configuration ID** to the filter key value we copied when we setup LTI Store. eg., `lti_store:1`
14. Click Save. 
15. The LTI Tool should load as expected.

### Additional Information

1. [ADR of the implementation](https://github.com/openedx/xblock-lti-consumer/blob/master/docs/decisions/0006-pluggable-lti-configuration.rst)
2. [Compatible Openedx Plugin app - openedx-ltistore](https://github.com/open-craft/openedx-ltistore)
